multicluster/controllers/multicluster/clusterset_controller.go

```diff
@@ -18,38 +18,93 @@ package multicluster
 
 import (
 	"context"
+	"fmt"
+	"sync"
 
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+	"antrea.io/antrea/multicluster/controllers/multicluster/core"
 )
 
 // ClusterSetReconciler reconciles a ClusterSet object
+// There will be one controller running in each namespace of the leader for multiple cluster-set support
+// so each controller will only be handling a single cluster set in the given namespace.
+// TODO: Split the reconciler for member and leader to avoid if checks. In case a cluster is both a
+//       member and a leader, 2 instances of the controller will need to be run, one as a leader
+//		 the second as a member.
 type ClusterSetReconciler struct {
 	client.Client
-	Scheme *runtime.Scheme
+	mutex    sync.Mutex
+	Scheme   *runtime.Scheme
+	IsLeader bool
+
+	clusterSetConfig *multiclusterv1alpha1.ClusterSet
+	clusterSetID     common.ClusterSetID
+	clusterID        common.ClusterID
+
+	// These fields are only applicable on leader cluster
+	LocalClusterManager core.LocalClusterManager
+
+	// These fields are only applicable on member cluster
+	RemoteClusterManager core.RemoteClusterManager
 }
 
 //+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the ClusterSet object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+// Reconcile ClusterSet CRD changes
 func (r *ClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	clusterSet := &multiclusterv1alpha1.ClusterSet{}
+	err := r.Get(ctx, req.NamespacedName, clusterSet)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		// if errors.IsNotFound(err)
+		klog.Infof("Received ClusterSet delete", "config", clusterSet, "leader", r.IsLeader)
+		if !r.IsLeader {
+			if err := r.RemoteClusterManager.Stop(); err != nil {
```

Comment:
I got below error when I am trying to delete a clusterset resource.:
```
I1122 07:47:47.275501       1 clusterset_controller.go:74] Received ClusterSet delete%!(EXTRA string=config, *v1alpha1.ClusterSet=&{{ } {      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[] [] } {0 0 [] [] 0}}, string=leader, bool=false)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x16fc74e]

goroutine 470 [running]:
antrea.io/antrea/multicluster/controllers/multicluster.(*ClusterSetReconciler).Reconcile(0xc000435730, {0x1dbaeb8, 0xc000b1ce10}, {{{0xc000f87680, 0x1a04ba0}, {0xc000f87670, 0xc000aa2000}}})
	/Users/luolan/workspaces/codes/src/github.com/luolanzone/antrea/multicluster/controllers/multicluster/clusterset_controller.go:76 +0x22e
```

Reply:
Ill try the delete workflow